aten/src/ATen/native/cuda/SpectralOps.cu

```diff
@@ -22,116 +25,110 @@ namespace at { namespace native {
 
 using namespace at::native::detail;
 
-// In real-to-complex transform, cuFFT only fills half of the values due to
-// conjugate symmetry. See native/SpectralUtils.h for more details.
-// The following structs are used to fill in the other half with symmetry in
-// case of real-to-complex transform with onesided=False flag.
-// See NOTE [ Fourier Transform Conjugate Symmetry ] in native/SpectralOpsUtils.h.
-
-// counting_iterator => index to fill
-struct cnt_to_dst_idx_functor : public thrust::unary_function<int64_t, int64_t>
-{
-  int64_t last_dim_size;
-  int64_t last_dim_start_slice;
-  int64_t last_dim_to_fill_size;
-
-  cnt_to_dst_idx_functor(int64_t last_dim_size, int64_t last_dim_start_slice) :
-    last_dim_size(last_dim_size), last_dim_start_slice(last_dim_start_slice),
-    last_dim_to_fill_size(last_dim_size - last_dim_start_slice) {}
-
-  // HIP wants __host__ __device__ tag, CUDA does not
-#ifdef __HIP_PLATFORM_HCC__
-  __host__ __device__
-#endif
-  cnt_to_dst_idx_functor & operator=(const cnt_to_dst_idx_functor&) = default;
-
-  __host__ __device__ __forceinline__
-  int64_t operator()(const int64_t& i) const
-  {
-    int64_t imag = i % 2;
-    int64_t idx = i / 2;
-    int64_t num_dim = idx / last_dim_to_fill_size;
-    int64_t slice_idx = idx % last_dim_to_fill_size;
-    return (num_dim * last_dim_size + last_dim_start_slice + slice_idx) * 2 + imag;
-  }
-};
+// Offset calculator for indexing in Hermitian mirrored order.
+// In mirrored dims, maps linear index i to (n - i) % n
+template <typename index_t>
+struct HermitianSymmetryOffsetCalculator {
+  using offset_type = at::detail::Array<index_t, 1>;
+  int64_t dims;
+  IntDivider<index_t> sizes_[MAX_DIMS];
+  index_t strides_[MAX_DIMS];
+  uint32_t mirror_dim_;  // bit mask
+  static_assert(MAX_DIMS < 32, "Need a bigger mask type");
+
+  HermitianSymmetryOffsetCalculator(
+      IntArrayRef sizes, IntArrayRef strides, IntArrayRef dim,
+      const int64_t element_size){
+    dims = sizes.size();
+    TORCH_INTERNAL_ASSERT(sizes.size() == strides.size());
+    TORCH_INTERNAL_ASSERT(dims <= MAX_DIMS);
+
+    for (int i = 0; i < MAX_DIMS; ++i) {
```

Comment:
Don't use an imprecise type like int. In this case you probably want to type i like MAX_DIMS:

```
auto i = decltype(MAX_DIMS){0}
```